text/0000-process-stdio-redirection.md

```diff
@@ -0,0 +1,161 @@
+OS pipe) without forcing the parent process to buffer the data itself.
+
+For example, one may wish to spawn a process which prints gigabytes
+of data (e.g. logs) and use another process to filter through it, and save the
+result to a file. The current API would force the parent process to dedicate a
+thread just to buffer all data between child processes and disk, which is
+impractical given that the OS can stream the data for us via pipes and file
+redirection.
+
+# Detailed design
+
+The least intrusive way to implement this functionality would be to implement
+`FromRaw{Fd, Handle}` for `Stdio`, taking ownership of the underlying handle.
+This approach would not require the addition of any new public APIs.
+
+There are several disadvantages to this design: it does not allow for using
+one handle for several redirections without forcing the caller to `unsafe`ly
+duplicate the handle, in addition, any cross-platform code will need to be
+littered with `cfg!` to call the appropriate methods. Furthermore, trying to use
+standard types like `File` will require that they be leaked by `mem::forget` so
+that their own destructors do not close the underlying handles. However, these
+inefficiencies can be addressed through a cross-platform, high level API.
+
+Lastly, the `AsRaw{Fd, Handle}` traits should be implemented for the
+`ChildStd{in,out,err}` types. This would allow easily piping output from one
+child process to another by leveraging the underlying OS pipes that were already
+created when spawning the child.
+
+## A High Level API Proposal
+
+In order to facilitate cross-platform usage, the API should be defined to
+operate on the respective `AsRaw{Fd, Handle}` traits for Unix and Windows,
+respectively. All method signatures should match lexically such that
+using `cfg!($platform)` checks will not be necessary. The private `StdioImp`
+enum in `std::process` should be extended with a `Redirect` variant which will
+hold a wrapper over the appropriate OS handle type to ensure it is properly
+closed (e.g. using an `AnonPipe`).
+
+Next, the API should expose methods for redirection that both do and do not take
+ownership of the underlying handle, e.g. `redirect<T: AsRaw*>(t: T)` and
+`redirect_by_ref<T: AsRaw*>(t: &T)`.
+
+The method that takes ownership retains the
+benefits of using `FromRaw*` directly while helping the caller avoid making
+platform specific calls. Unfortunately, since we cannot guarantee that an
+implementor of `AsRaw*` is the sole owner of the OS handle they return, this
+method will have to be `unsafe`.
+
+The method which does not take ownership allows the
+caller to reuse their handle without making excessive duplications, which would
+not be possible by using `FromRaw*` directly. The caller is, however, forced to
+ensure the handle will remain valid until the child is spawned, making this
+method `unsafe` as well.
+
+Below are several alternative ways of exposing a high level API. They are
```

Comment:
I would recommend structuring this RFC to make a specific recommendation on which API route to choose and leave the other two to the "Alternatives" section. I would personally recommend something like:

1. Having an extra `StdioExt` struct seems somewhat odd as it's not actually returned or needed to be a struct, so I'd list this as an alternative.
2. Using an extension trait is currently the standard method for extending functionality on a type, so I'd recommend listing this as the what the RFC proposes.
3. I would recommend listing this as an alternative, but definitely emphasize the non-cross-platform aspect as it's the reason I wouldn't want to pursue it.